scvi/external/scbasset/_model.py

```diff
@@ -192,6 +205,191 @@ def get_cell_bias(self) -> np.ndarray:
+
+        Parameters
+        ----------
+        tf
+            transcription factor name. must be provided in the relevant motif repository.
+        genome
+            species name for the motif injection procedure. Currently, only "human"
+            is supported.
+        motif_dir
+            path for the motif library. Will download if not already present.
+        lib_size_norm
+            normalize accessibility scores for library size by *substracting* the
+            cell bias term from each accessibility score prior to comparing motif
+            scores to background scores.
+        batch_size
+            minibatch size for TF activity inference.
+
+        Returns
+        -------
+        tf_score
+            [cells,] TF activity scores.
+
+        Notes
+        -----
+        scBasset infers TF activities by injecting known TF motifs into a
+        shuffled dinucleotide sequence and computing the change in accessibility
+        predicted between the injected motif and a randomized background
+        sequence. See :cite:p:`Yuan2022` for details. We modeled this function
+        off the original implementation in `scbasset`.
+
+        https://github.com/calico/scBasset/blob/9a3294c240d849cdac62682e324bc5f4836bb744/scbasset/utils.py#L453
+        """
+        # check for a library of FASTA sequences corresponding to motifs and
+        # download if none is found
+        # `motif_seqs` is a List of str sequences where each char is in "ACTGN".
+        # `bg_seqs` is the same, but background sequences rather than motif injected
+        motif_seqs, bg_seqs = self._get_motif_library(
+            tf=tf, genome=genome, motif_dir=motif_dir
+        )
+
+        # SCBASSET.module.inference(...) takes `dna_code: torch.Tensor` as input
+        # where `dna_code` is [batch_size, seq_length] and each value is [0,1,2,3]
+        # where [0: A, 1: C, 2: G, 3: T].
+        motif_codes = pd.DataFrame([list(s) for s in motif_seqs]).applymap(_dna_to_code)
+        bg_codes = pd.DataFrame([list(s) for s in bg_seqs]).applymap(_dna_to_code)
+        # [batch_size, seq_length]
+        motif_codes = torch.from_numpy(np.array(motif_codes)).long()
+        bg_codes = torch.from_numpy(np.array(bg_codes)).long()
+
+        # NOTE: The below modification is added due to a difference between the original
+        # and `scvi-tools` implementation of scBasset.
+        # `scvi-tools` uses a fixed length of 1334, but the original model used a fixed length of 1344.
+        # This is because the `MaxPool1d` operations in the `scvi-tools` implementation
+        # use `ceil_mode=True` by default, yielding the expected bottleneck shape of
+        # (batch_size, 7) for inputs of 1334, NOT 1344. If `ceil_mode=False` is used instead
+        # we would reproduce the original scBasset paper.
+        #
+        # Here, we crop the motifs to make the original libraries compatabible with the
+        # `scvi-tools` implementation.
```

Comment:
Actually, last thing, this isn't applicable anymore right?

Reply:
Oh yeah, missed this one. Replaced with a warning to users if they decide to load motifs from another source that get trimmed, throws `ValueError` if motif sizes are smaller than the input size of the model.